### PR TITLE
Shape boolean operation improvements

### PR DESCRIPTION
**Pull request by caleb-ad**
- [x] Handle overlapping curve case
- [x] Implement boolean operations for selections of many shapes, including proper difference
- [x] Add proper undo behavior
- [x] Handle the ray-cast-into-endpoint edge case more elegantly
- [x] Further improvements and bugfixes

### CODE REVIEW

**Comment by Keavon** [comment]
Checking in on the progress with this.

**Comment by Keavon** [comment]
@caleb-ad could this be finished up soon?

**Comment by caleb-ad** [comment]
@Keavon Yes, I some changes which are nearly finished which should wrap this up. I think i'll have that done by friday.

**Comment by Keavon** [comment]
Glad to hear it! If you have trouble merging (and we have a lot of PRs going in right now which you'll have to keep merging in), please reach out and I or @0HyperCube would be happy to help with the merge conflict.

**Comment by Keavon** [comment]
So it's all ready do code review and merge at this moment?

File: graphene/src/boolean_ops.rs [hunk 1]
```diff
@@ -55,7 +48,17 @@ struct Vertex {
 
 impl Debug for Vertex {
 	fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-		f.write_str(format!("\n    Intersect@ {:?}", self.intersect.point).as_str())?;
+		f.write_str(
+			format!(
+				"\n    Intersect@ {:?} a_seg: {:?}, b_seg: {:?} t_a: {:?} t_b: {:?}",
```
Comment (Keavon) [inline graphene/src/boolean_ops.rs:53]: - Why the 4 spaces at the start of the string? If they do in fact belong, adding a comment explaining this question would be helpful.
- Is that at sign supposed to be a colon in `Intersect@`?
- the `_seg` and `t_` suffixes and prefixes are pretty cryptic to me. Could you pick full words that describe exactly what these refer to?
Thanks!

File: graphene/src/boolean_ops.rs [hunk 2]
```diff
@@ -453,12 +473,77 @@ pub fn subdivide_path_seg(p: &PathSeg, t_values: &mut [f64]) -> Vec<Option<PathS
 	sub_segments
 }
 
+/// * shapes.len() > 1
```
Comment (Keavon) [inline graphene/src/boolean_ops.rs:476]: This doc comment is meant to describe the functionality of this function, but the comment doesn't really make any sense to me.

**Review by Keavon — APPROVED** [review]

**Pull request merged**